rest_framework/schemas.py

```diff
@@ -256,6 +256,288 @@ def get_allowed_methods(self, callback):
+
+        if not hasattr(view, 'get_serializer'):
+            return []
+
+        serializer = view.get_serializer()
+
+        if isinstance(serializer, serializers.ListSerializer):
+            return [
+                coreapi.Field(
+                    name='data',
+                    location='body',
+                    required=True,
+                    schema=coreschema.Array()
+                )
+            ]
+
+        if not isinstance(serializer, serializers.Serializer):
+            return []
+
+        fields = []
+        for field in serializer.fields.values():
+            if field.read_only or isinstance(field, serializers.HiddenField):
+                continue
+
+            required = field.required and method != 'PATCH'
+            field = coreapi.Field(
+                name=field.field_name,
+                location='form',
+                required=required,
+                schema=field_to_schema(field)
+            )
+            fields.append(field)
+
+        return fields
+
+    def get_pagination_fields(self, path, method):
+        view = self.view
+
+        if not is_list_view(path, method, view):
+            return []
+
+        pagination = getattr(view, 'pagination_class', None)
+        if not pagination:
+            return []
+
+        paginator = view.pagination_class()
+        return paginator.get_schema_fields(view)
+
+    def get_filter_fields(self, path, method):
+        view = self.view
+
+        if not is_list_view(path, method, view):
+            return []
+
+        if not getattr(view, 'filter_backends', None):
+            return []
+
+        fields = []
+        for filter_backend in view.filter_backends:
+            fields += filter_backend().get_schema_fields(view)
+        return fields
+
+    def get_encoding(self, path, method):
+        """
+        Return the 'encoding' parameter to use for a given endpoint.
+        """
+        view = self.view
+
+        # Core API supports the following request encodings over HTTP...
+        supported_media_types = set((
+            'application/json',
+            'application/x-www-form-urlencoded',
+            'multipart/form-data',
+        ))
+        parser_classes = getattr(view, 'parser_classes', [])
+        for parser_class in parser_classes:
+            media_type = getattr(parser_class, 'media_type', None)
+            if media_type in supported_media_types:
+                return media_type
+            # Raw binary uploads are supported with "application/octet-stream"
+            if media_type == '*/*':
+                return 'application/octet-stream'
+
+        return None
+
+# TODO: Where should this live?
+#   - We import APIView here. So we can't import the descriptor into `views`
+#   - APIView is only used by SchemaView.
+#       - ???: Make `schemas` a package and move SchemaView to `schema.views`
+#       - That way the schema attribute could be set in the class definition.
+APIView.schema = AutoSchema()
```

Comment:
@tomchristie What do you think about this one? 

If we adjust the dependency graph needs to go like this: 

```
┌────────────┐       ┌────────┐   
│  schemas   │──────▶│ views  │   
└────────────┘       └────────┘   
       │                  │       
       │                  │       
       │                  ▼       
       │          ┌──────────────┐
       └─────────▶│schemas.views │
                  └──────────────┘

```

Reply:
Options: 

1. Leave it like this. It's a _monkey patch_ — but all the schema stuff stays in one file
2. Make schemas a package: 
  * More complex 
  * But keeps `APIView` totally declarative.

Reply:
I think I prefer Option 1.